File: common/src/main/java/com/pluxity/cctv/Cctv.java
package com.pluxity.cctv;

import com.pluxity.cctv.category.CctvCategory;
import com.pluxity.cctv.dto.CctvUpdateRequest;
import com.pluxity.feature.entity.Feature;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Cctv {
    @Id private String id;
    private String name;

    @Column(length = 1000)
    private String url;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "feature_id")
    private Feature feature;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private CctvCategory category;

    @Builder
    public Cctv(String id, String name, String url, CctvCategory category, Feature feature) {
        this.id = id;
        this.name = name;
        this.url = url;
        this.category = category;
        this.feature = feature;
    }

    public void updateCctv(CctvUpdateRequest request) {
        this.name = request.name();
        this.url = request.url();
    }

    public void changeFeature(Feature feature) {
        this.feature = feature;
    }

    public void changeCategory(CctvCategory category) {
        this.category = category;
    }
}

File: gs/src/main/kotlin/com/pluxity/cctv/entity/Cctv.kt
package com.pluxity.cctv.entity

import com.pluxity.cctv.dto.CctvUpdateRequest
import com.pluxity.feature.entity.Feature
import com.pluxity.global.entity.BaseEntity
import com.pluxity.permission.ResourceType
import com.pluxity.user.entity.Permissible
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.OneToOne

@Entity
class Cctv(
    @Id
    var id: String,
    // DB 컬럼은 name 유지
    var name: String = "",
    @Column(length = 1000)
    var url: String?,
    @OneToOne(fetch = FetchType.LAZY, cascade = [CascadeType.PERSIST])
    @JoinColumn(name = "feature_id")
    var feature: Feature? = null,
) : BaseEntity(),
    Permissible {
    fun updateCctv(request: CctvUpdateRequest) {
        this.name = request.name
        this.url = request.url
    }

    fun changeFeature(feature: Feature?) {
        this.feature = feature
    }

    override val resourceId: String
        get() = id

    override val resourceType: ResourceType
        get() = ResourceType.CCTV
}
